ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveRowIsDeletedPropagatorRule.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.rules.views;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.hadoop.hive.ql.optimizer.calcite.rules.HiveFieldTrimmerRule;
+
+/**
+ * Planner rule to propagate rowIsDeleted column in Materialized View incremental rebuild plans.
+ * This rule uses {@link HiveRowIsDeletedPropagator}.
+ * This rule should be applied on the uppermost Right outer join node.
+ */
+public class HiveRowIsDeletedPropagatorRule extends HiveFieldTrimmerRule {
```

Comment:
Is it an overkill that this rule extends `HiveFieldTrimmerRule` since it is unrelated? Should they both extend a common class that has the mechanism to unnest the plan and trigger a single time instead?

I think an alternative that could be much simpler is to use the `HiveRowIsDeletedPropagator` uniquely (no rule) since it already implements a traversal logic. You could control the flow internally with a boolean that indicates whether you have already passed through the RIGHT join while doing the traversal. That would also avoid any changes in `HiveFieldTrimmerRule` and simplify the patch.

Reply:
I went with the no rule approach. This case a new `RelBuilder` instance is required: 
```
basePlan = new HiveRowIsDeletedPropagator(HiveRelFactories.HIVE_BUILDER.create(this.cluster, null))
     .propagate(basePlan);
```